pkg/agent/controller/packetsampling/packetsampling_controller.go

```diff
@@ -0,0 +1,821 @@
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing PacketSampling for %s. (%v)", psName, time.Since(startTime))
+	}()
+
+	ps, err := c.packetSamplingLister.Get(psName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			c.cleanupPacketSampling(psName)
+			return nil
+		}
+		return err
+	}
+
+	switch ps.Status.Phase {
+	case crdv1alpha1.PacketSamplingRunning:
+		if ps.Status.DataplaneTag != 0 {
+			start := false
+			c.runningPacketSamplingsMutex.Lock()
+			if _, ok := c.runningPacketSamplings[ps.Status.DataplaneTag]; !ok {
+				start = true
+			}
+			c.runningPacketSamplingsMutex.Unlock()
+			if start {
+				err = c.startPacketSampling(ps)
+			}
+		} else {
+			klog.Warningf("Invalid data plane tag %d for packet %s", ps.Status.DataplaneTag, ps.Name)
+		}
+	default:
+		c.cleanupPacketSampling(psName)
+	}
+	return err
+
+}
+
+type uploader interface {
+	upload(addr string, path string, config *ssh.ClientConfig, tarGzFile io.Reader) error
+}
+
+type sftpUploader struct {
+}
+
+func (uploader *sftpUploader) upload(address string, path string, config *ssh.ClientConfig, packetFile io.Reader) error {
+	conn, err := ssh.Dial("tcp", address, config)
+	if err != nil {
+		return fmt.Errorf("error when connecting to fs server: %w", err)
+	}
+	sftpClient, err := sftp.NewClient(conn)
+	if err != nil {
+		return fmt.Errorf("error when setting up sftp client: %w", err)
+	}
+	defer func() {
+		if err := sftpClient.Close(); err != nil {
+			klog.ErrorS(err, "Error when closing sftp client")
+		}
+	}()
+	targetFile, err := sftpClient.Create(path)
+	if err != nil {
+		return fmt.Errorf("error when creating target file on remote: %v", err)
+	}
+	defer func() {
+		if err := targetFile.Close(); err != nil {
+			klog.ErrorS(err, "Error when closing target file on remote")
+		}
+	}()
+	if written, err := io.Copy(targetFile, packetFile); err != nil {
+		return fmt.Errorf("error when copying target file: %v, written: %d", err, written)
+	}
+	klog.InfoS("Successfully upload file to path", "filePath", path)
+	return nil
+}
+
+func (c *Controller) getUploaderByProtocol(protocol StorageProtocolType) (uploader, error) {
+	if protocol == sftpProtocol {
+		return c.sftpUploader, nil
+	}
+	return nil, fmt.Errorf("unsupported protocol %s", protocol)
+}
+
+func parseUploadUrl(uploadUrl string) (*url.URL, error) {
+	parsedURL, err := url.Parse(uploadUrl)
+	if err != nil {
+		parsedURL, err = url.Parse("sftp://" + uploadUrl)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if parsedURL.Scheme != "sftp" {
+		return nil, fmt.Errorf("not sftp protocol")
+	}
+	return parsedURL, nil
+}
+
+// parseBundleAuth returns the authentication from the Secret provided in BundleServerAuthConfiguration.
+// The authentication is stored in the Secret Data with a key decided by the AuthType, and encoded using base64.
+func (c *Controller) parseBundleAuth(authentication crdv1alpha1.BundleServerAuthConfiguration) (*controlplane.BundleServerAuthConfiguration, error) {
```

Comment:
It looks the same as the one in supportbundlecollection's controller, could you check and refine this as a utility  function in a shared file? maybe add a new package 'uploader' in the util folder and move related functions into this folder.

Reply:
Yes, i have create a new package in util named `ftp` to provide common interface for packetsampling and supportbundlecollection.